libfwupdplugin/fu-bluez-device.c

```diff
@@ -293,59 +291,151 @@ fu_bluez_device_get_ble_string_property(const gchar *obj_path,
 	return g_variant_dup_string(val, NULL);
 }
 
+static gchar *
+fu_bluez_device_get_interface_uuid(FuBluezDevice *self,
+				   GDBusObject *obj,
+				   const gchar *obj_path,
+				   const gchar *iface_name)
+{
+	g_autoptr(GDBusInterface) iface = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *obj_uuid = NULL;
+
+	iface = g_dbus_object_get_interface(obj, iface_name);
+	if (iface == NULL)
+		return NULL;
+	obj_uuid =
+	    fu_bluez_device_get_ble_string_property(obj_path, iface_name, "UUID", &error_local);
+	if (obj_uuid == NULL)
+		g_debug("failed to get %s property: %s", iface_name, error_local->message);
+
+	return g_steal_pointer(&obj_uuid);
+}
+
+/*
+ * Populates the {uuid_helper : object_path} entry of a device for its
+ * characteristic.
+ */
+static gboolean
+fu_bluez_device_add_characteristic_uuid(FuBluezDevice *self,
+					GDBusObject *obj,
+					const gchar *obj_path,
+					const gchar *iface_name)
+{
+	g_autofree gchar *obj_uuid =
+	    fu_bluez_device_get_interface_uuid(self, obj, obj_path, iface_name);
+
+	if (obj_uuid == NULL)
+		return FALSE;
+
+	fu_bluez_device_add_uuid_path(self, obj_uuid, obj_path);
+	return TRUE;
+}
+
+static gboolean
+fu_bluez_device_add_instance_by_service_uuid(FuBluezDevice *self,
+					     GDBusObject *obj,
+					     const gchar *obj_path,
+					     const gchar *iface_name)
+{
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *obj_uuid =
+	    fu_bluez_device_get_interface_uuid(self, obj, obj_path, iface_name);
+
+	if (obj_uuid == NULL)
+		return FALSE;
+
+	/* Register device by service UUID. */
+	fu_device_add_instance_str(FU_DEVICE(self), "GATT", obj_uuid);
+	if (!fu_device_build_instance_id_full(FU_DEVICE(self),
+					      FU_DEVICE_INSTANCE_FLAG_VISIBLE |
+						  FU_DEVICE_INSTANCE_FLAG_QUIRKS,
+					      &error_local,
+					      "BLUETOOTH",
+					      "GATT",
+					      NULL)) {
+		g_debug("failed to register %s service: %s", obj_uuid, error_local->message);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_bluez_device_read_battery_interface(FuBluezDevice *self,
+				       GDBusObject *obj,
+				       const gchar *obj_path,
+				       const gchar *iface_name)
+{
+	guint8 percentage = FWUPD_BATTERY_LEVEL_INVALID;
+	g_autoptr(GDBusInterface) iface = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GVariant) obj_percentage = NULL;
+
+	iface = g_dbus_object_get_interface(obj, iface_name);
+	if (iface == NULL)
+		return FALSE;
```

Comment:
set a GError here

Reply:
the same -- what should we do with this error?
Interface may be absent, and this is a normal situation.

Reply:
yes, that's fine -- but that's a decision I think the caller of fu_bluez_device_read_battery_interface() should make -- i.e. it's fine to ignore but g_debug() the error in fu_bluez_device_ensure_gatt_interfaces() -- and set a GError literal like "no battery interface" here.